File: app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/AttachPhotoFragment.java
package de.westnordost.streetcomplete.quests.note_discussion;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.v4.app.Fragment;
import android.support.v4.content.FileProvider;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;

import de.westnordost.streetcomplete.ApplicationConstants;
import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.osmnotes.AttachPhotoUtils;

import static android.app.Activity.RESULT_OK;

public class AttachPhotoFragment extends Fragment
{
	private ArrayList<String> imagePaths;
	private String currentImagePath;

	private NoteImageAdapter noteImageAdapter;

	private static final String TAG = "AttachPhotoFragment";
	private static final int REQUEST_TAKE_PHOTO = 1;

	private static final String PHOTO_PATHS = "photo_paths";
	private static final String CURRENT_PHOTO_PATH = "current_photo_path";

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState)
	{
		View view = inflater.inflate(R.layout.fragment_attach_photo, container, false);

		ImageButton takePhoto = view.findViewById(R.id.buttonTakeImage);
		takePhoto.setOnClickListener(v -> takePhoto());

		if (!getActivity().getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA))
		{
			view.setVisibility(View.GONE);
		}

		if (savedInstanceState != null)
		{
			imagePaths = savedInstanceState.getStringArrayList(PHOTO_PATHS);
			currentImagePath = savedInstanceState.getString(CURRENT_PHOTO_PATH);
		}
		else
		{
			imagePaths = new ArrayList<>();
			currentImagePath = null;
		}

		noteImageAdapter = new NoteImageAdapter(imagePaths, getContext());
		RecyclerView gridView = view.findViewById(R.id.gridView);
		gridView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
		gridView.setAdapter(noteImageAdapter);

		return view;
	}

	@Override public void onSaveInstanceState(Bundle outState)
	{
		super.onSaveInstanceState(outState);
		outState.putStringArrayList(PHOTO_PATHS, imagePaths);
		outState.putString(CURRENT_PHOTO_PATH, currentImagePath);
	}

	private void takePhoto()
	{
		Intent takePhotoIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
		if (takePhotoIntent.resolveActivity(getActivity().getPackageManager()) != null)
		{
			try
			{
				File photoFile = createImageFile();
				Uri photoUri;
				if (Build.VERSION.SDK_INT > 21)
				{
					//Use FileProvider for getting the content:// URI, see: https://developer.android.com/training/camera/photobasics.html#TaskPath
					photoUri = FileProvider.getUriForFile(getActivity(), getString(R.string.fileprovider_authority), photoFile);
				}
				else
				{
					photoUri = Uri.fromFile(photoFile);
				}
				currentImagePath = photoFile.getPath();
				takePhotoIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoUri);
				startActivityForResult(takePhotoIntent, REQUEST_TAKE_PHOTO);
			}
			catch (IOException | IllegalArgumentException e)
			{
				Log.e(TAG, "Unable to create file for photo", e);
				Toast.makeText(getContext(), R.string.quest_leave_new_note_create_image_error, Toast.LENGTH_SHORT).show();
			}
		}
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, Intent data)
	{
		if (requestCode == REQUEST_TAKE_PHOTO)
		{
			if (resultCode == RESULT_OK)
			{
				try
				{
					Bitmap bitmap = AttachPhotoUtils.resize(currentImagePath, ApplicationConstants.ATTACH_PHOTO_MAXWIDTH);
					if(bitmap == null) throw new IOException();
					FileOutputStream out = new FileOutputStream(currentImagePath);
					bitmap.compress(Bitmap.CompressFormat.JPEG, ApplicationConstants.ATTACH_PHOTO_QUALITY, out);

					imagePaths.add(currentImagePath);
					noteImageAdapter.notifyItemInserted(imagePaths.size()-1);
				}
				catch (IOException e)
				{
					Log.e(TAG, "Unable to rescale the photo", e);
					Toast.makeText(getContext(), R.string.quest_leave_new_note_create_image_error, Toast.LENGTH_SHORT).show();
					removeCurrentImage();
				}
			}
			else
			{
				removeCurrentImage();
			}
			currentImagePath = null;
		}
	}

	private void removeCurrentImage()
	{
		if(currentImagePath != null)
		{
			File photoFile = new File(currentImagePath);
			if (photoFile.exists())
			{
				photoFile.delete();
			}
		}
	}

	private File createImageFile() throws IOException
	{
		File directory = getActivity().getExternalFilesDir(Environment.DIRECTORY_PICTURES);
		return File.createTempFile("photo", ".jpg", directory);
	}

	public ArrayList<String> getImagePaths()
	{
		return imagePaths;
	}

	public void deleteImages()
	{
		AttachPhotoUtils.deleteImages(imagePaths);
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/note_discussion/AttachPhotoFragment.kt
package de.westnordost.streetcomplete.quests.note_discussion

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.pm.PackageManager.FEATURE_CAMERA_ANY
import android.graphics.Bitmap
import android.os.Bundle
import android.os.Environment
import android.view.View
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.FileProvider
import androidx.core.view.isGone
import androidx.exifinterface.media.ExifInterface
import androidx.exifinterface.media.ExifInterface.TAG_GPS_IMG_DIRECTION
import androidx.exifinterface.media.ExifInterface.TAG_GPS_IMG_DIRECTION_REF
import androidx.fragment.app.Fragment
import de.westnordost.streetcomplete.ApplicationConstants
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osmnotes.deleteImages
import de.westnordost.streetcomplete.databinding.FragmentAttachPhotoBinding
import de.westnordost.streetcomplete.util.decodeScaledBitmapAndNormalize
import de.westnordost.streetcomplete.util.ktx.nowAsEpochMilliseconds
import de.westnordost.streetcomplete.util.ktx.toast
import de.westnordost.streetcomplete.util.logs.Log
import de.westnordost.streetcomplete.util.viewBinding
import de.westnordost.streetcomplete.view.AdapterDataChangedWatcher
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.lang.Exception

class AttachPhotoFragment : Fragment(R.layout.fragment_attach_photo) {

    private val binding by viewBinding(FragmentAttachPhotoBinding::bind)

    private val takePhoto = registerForActivityResult(ActivityResultContracts.TakePicture(), ::onTookPhoto)

    private lateinit var noteImageAdapter: NoteImageAdapter
    private var newPhotoFile: File? = null

    val imagePaths: List<String> get() = noteImageAdapter.list
    var hasGpxAttached = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val paths = savedInstanceState?.getStringArrayList(PHOTO_PATHS) ?: ArrayList()
        noteImageAdapter = NoteImageAdapter(paths, requireContext())
        newPhotoFile = savedInstanceState?.getString(NEW_PHOTO_FILE)?.let { File(it) }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        view.isGone = !requireActivity().packageManager.hasSystemFeature(FEATURE_CAMERA_ANY)

        binding.takePhotoButton.setOnClickListener { takePhoto() }
        binding.photosList.adapter = noteImageAdapter
        noteImageAdapter.registerAdapterDataObserver(AdapterDataChangedWatcher { updateHintVisibility() })

        binding.attachedGpxView.isGone = !hasGpxAttached

        updateHintVisibility()
    }

    private fun updateHintVisibility() {
        binding.photosList.isGone = imagePaths.isEmpty()
        binding.photosAreUsefulExplanation.isGone = imagePaths.isNotEmpty()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putStringArrayList(PHOTO_PATHS, ArrayList(imagePaths))
        outState.putString(NEW_PHOTO_FILE, newPhotoFile?.path)
    }

    private fun takePhoto() {
        val activity = activity ?: return
        try {
            val file = createImageFile(activity)
            newPhotoFile = file
            val photoUri = FileProvider.getUriForFile(activity, activity.getString(R.string.fileprovider_authority), file)
            takePhoto.launch(photoUri)
        } catch (e: Exception) {
            newPhotoFile?.let { deleteImageFile(it) }
            when (e) {
                is ActivityNotFoundException -> context?.toast(R.string.no_camera_app)
                else -> {
                    Log.e(TAG, "Unable to create photo", e)
                    context?.toast(R.string.quest_leave_new_note_create_image_error)
                }
            }
        }
    }

    private fun onTookPhoto(hasSavedPhoto: Boolean) {
        val file = newPhotoFile
        if (!hasSavedPhoto) {
            newPhotoFile?.let { deleteImageFile(it) }
            return
        }
        if (file != null) {
            val exif = ExifInterface(file)
            rescaleImageFile(file)
            copyDirectionExifData(file, exif)

            noteImageAdapter.list.add(file.path)
            noteImageAdapter.notifyItemInserted(imagePaths.size - 1)
        }
    }

    fun deleteImages() {
        deleteImages(imagePaths)
    }

    companion object {
        private const val TAG = "AttachPhotoFragment"
        private const val PHOTO_PATHS = "photo_paths"
        private const val NEW_PHOTO_FILE = "photo_file"
    }
}

private fun createImageFile(context: Context): File {
    val directory = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
    val imageFileName = "photo_" + nowAsEpochMilliseconds() + ".jpg"
    val file = File(directory, imageFileName)
    if (!file.createNewFile()) throw IOException("Photo file with exactly the same name already exists")
    return file
}

private fun rescaleImageFile(file: File) {
    val bitmap = decodeScaledBitmapAndNormalize(file.path, ApplicationConstants.ATTACH_PHOTO_MAXWIDTH, ApplicationConstants.ATTACH_PHOTO_MAXHEIGHT) ?: throw IOException()
    val out = FileOutputStream(file.path)
    bitmap.compress(Bitmap.CompressFormat.JPEG, ApplicationConstants.ATTACH_PHOTO_QUALITY, out)
}

private fun copyDirectionExifData(file: File, exif: ExifInterface) {
    val newExif = ExifInterface(file.path)
    newExif.setAttribute(TAG_GPS_IMG_DIRECTION, exif.getAttribute(TAG_GPS_IMG_DIRECTION))
    newExif.setAttribute(TAG_GPS_IMG_DIRECTION_REF, exif.getAttribute(TAG_GPS_IMG_DIRECTION_REF))
    newExif.saveAttributes()
}

private fun deleteImageFile(file: File) {
    if (file.exists()) file.delete()
}
